How to create GIN private repo for crab data (#81)
SainsburyWellcomeCentre/crabs-exploration: guides/CreateGINPrivateRepo.md

# Create a GIN private repository

> GIN is a free and open data management system designed for reproducible management of neuroscientific data. It is a web-accessible repository store of your data based on `git` and `git-annex` that you can access securely anywhere you desire while keeping your data in sync, backed up and easily accessible.

Reviewer: ```suggestion
[GIN](https://gin.g-node.org/G-Node/Info/wiki) (hosted by the German Neuroinformatics Node) is a free and open data management system designed for reproducible management of neuroscientific data. It is a web-accessible repository store of your data based on `git` and `git-annex` that you can access securely anywhere you desire while keeping your data in sync, backed up and easily accessible.
```

> Below we explain the steps to create a GIN private repository for a dataset (in our case, a set of labelled images). Note that all GIN repos are private by default.

Reviewer: ```suggestion
Below we explain the steps to create a GIN private repository for a dataset (in our case, a set of labelled images). 

> [!NOTE]
> All GIN repos are private by default.
```

   ```
   gin login
   ```

   - You will be prompted for your GIN username and password.
   - To list the repositories available to your account: `gin repos --all`

> [!TIP]
>
> You may need `sudo` permissions for some of the following `gin` commands. If so, remember to prepend all commands with `sudo`.

2. **Initialise a GIN repository**

   - **Option 1: on a new directory**

     - Create a new GIN repository locally and on the GIN server by running:

       ```
       gin create <repository name>
       ```

         <details><summary> OR alternatively:</summary>

       Create a repository in the GIN server [from the browser](https://gin.g-node.org/repo/create), and download it locally to your local workspace by running:

       ```
       gin get <user name>/<repository name>
       ```

         </details>

     - Once the repository has been initialised, add data to the new local GIN repository with `mv`, `cp` or drag-and-dropping files to the directory.

   - **Option 2: on an existing directory**

     - To create a new repository on the GIN server and in the current working directory in one go, run:

       ```
       gin create --here <name>
       ```

> 
>         <details><summary> OR alternatively:</summary>
> 
>        To do each step independently:

Reviewer: ```suggestion

        <details><summary>Or, to do each step independently:</summary>      
```

       ```
       gin add-remote <name> <location>
       ```

       where `<name>` is the name you want to give to the remote (e.g. `origin`) and `<location>` is the location of the data store, which should be in the form of alias:path or server:path (e.g. `gin add-remote origin gin:sfmig/crab-data`). - If the remote GIN repository doesn't exist, you will be prompted to either create the remote GIN repository, add the remote address anyways or abort. - To show the remotes accessible to your GIN account run `gin remotes`.
        </details>

> [!TIP]
>
> To create a GIN repository on a `ceph` directory:
>
> - You may need to mount the `ceph` directory first. To do this temporarily (i.e., until the next reboot), follow [this guide](https://howto.neuroinformatics.dev/programming/Mount-ceph-ubuntu-temp.html). To do this permanently, follow [this one](https://howto.neuroinformatics.dev/programming/Mount-ceph-ubuntu.html).
> - You may also need to add an exception for the mounted directory. To do so, run the following command:
>
>   ```
>   git config --global --add safe.directory /mnt/<path-to-the-mounted-directory>
>   ```
>

> > - Alternatively, you can log to SWC's HPC cluster (specifically, its [gateway node](https://howto.neuroinformatics.dev/_images/swc_hpc_access_flowchart.png) `hpc-gw1`), which has the GIN CLI client installed, and work from there. This is likely faster than mounting the `ceph` directory in your laptop, since the cluster is in the same network as `ceph` (and actually physically close to it).

Reviewer: 😎

   With the `--content` flag, it optionally downloads the content of all files in the repository. If `--content` is not specified, new files will be empty placeholders.

   Content of individual files can be retrieved using the `gin get-content <filename>` command, and later removed with `gin remove-content <filename>`. See [the docs](https://gin.g-node.org/G-Node/Info/wiki/GIN+CLI+Help#download-all-new-information-from-a-remote-repository) for further details.

To download the data programmatically in your Python code:

- We recommend using [pooch](https://www.fatiando.org/pooch/latest/index.html) to easily download data from the GIN repo's URL. Pooch also has some nice bonus functionalities like caching the downloaded data, verifying cryptographic hashes or unzipping files upon download.

## Other useful tips

- To [unannex a file](https://gin.g-node.org/G-Node/Info/wiki/FAQ+Troubleshooting#how-to-unannex-files), aka remove a file from the GIN tracking before uploading:

  ```

>   gin git annex unannex [path/filename]

Reviewer: `wrapper wrapped-thing do actually-undo`
🫠

  ```

- To stop tracking the GIN repo locally delete the `.git` directory

- To delete a GIN repository but keep the git repo:

  - delete the repository in the GIN server via the browser
  - delete the GIN local repository with `git annex uninit`
    - this command removes relevant bits in `.git/annex` and `.git/objects`, but some pre-commits may need to be edited by hand (see this [SO post](https://stackoverflow.com/questions/24447047/remove-git-annex-repository-from-file-tree)).

## File locking

> [File locking](https://gin.g-node.org/G-Node/Info/wiki/GIN+CLI+Usage+Tutorial#file-locking) is an important point in GIN repos and git-annex which surprisingly comes up quite late in the docs. Below the main ideas behind it.

Reviewer: ```suggestion
[File locking](https://gin.g-node.org/G-Node/Info/wiki/GIN+CLI+Usage+Tutorial#file-locking) is an important point in GIN repos and git-annex which surprisingly comes up quite late in the docs. Below, are the main ideas behind this.
```

- Files in a GIN repo can be locked or unlocked.

- The lock state of a file is persistent. This means that if I clone a GIN repo whose files are unlocked, I lock them in my local copy, and then upload that to the GIN server, the next time someone clones from the GIN repo the files they fetch will be locked.

> - The lock state relates to the nature of the placeholder file (aka, the file in the working directory when we do `gin get <repository`):

Reviewer: ```suggestion
- The lock state actually relates more to the nature of the placeholder file (the file in the working directory when we do `gin get <repository`):
```

  - on Unix-like systems: if a file is locked, its corresponding placeholder file will be a symlink pointing to the annexed content (i.e., the content under `.git/annex/objects`). This way we can open and inspect the file but not modify it. If a file is unlocked, the placeholder file in the working directory is an ASCII text file with a path. This path is _approximately_ where the content of the file will be downloaded to when we request it.
  - on Windows: from the docs, if a file is locked, the placeholder file is a plain text file pointing to the content in the git annex. If a file is unlocked, presumably the behaviour is the same as in Unix-like systems. I haven't tested the situation on Windows.

> - Unlocked files can be edited. If the data is unlocked and the full content of the dataset is downloaded locally, the file in the working directory has content, and so does its copy under git annex. This doubles disk usage of files checked into the repo, but in exchange users can modify and revert files to previous commits.
> 

Reviewer: ```suggestion
- Unlocked files can be edited. If the data is unlocked and the full content of the dataset is downloaded locally, the file in the working directory has content, and so does its copy under git annex. 
  > [!CAUTION]
  > This doubles disk usage of files checked into the repo, but in exchange users can modify and revert files to previous commits.

```
Not actually sure if you can have a GFM callout as a sub bullet bit if possible I'd highlight this in a box here. Feels like it's quite important to note that running something as naively simple as
```
gin unlock *
```
could write GBs of data.

> - Locked files cannot be edited. In MacOS, we may be prompted to unlock the files if we try to edit them, but we won't be able to save any changes because we don't have writing permissions. Files need to be committed before locking (double check).

Reviewer: I don't understand.

On MacOS who prompts you to unlock when you try to edit? Is it a GIN tool?

  - To get the actual content in the git annex, we need to run `gin download --content`. This will fetch the content from the GIN server (presumably). After this, the symlinks in the working directory should work

- And in an existing directory?

  - After initialising the GIN repo in the current directory and adding a remote, we would commit the data. When committing, the data is "copied" to the git annex.
  - To replace the copies in the working directory with symlinks to the git annex content, we run `gin lock <path-to-data>`
  - If we commit this state change and upload the changes to the GIN server, the files will be locked for any future clones of the repo.

- Useful tools for inspecting this further

  - `file` shows the type of file (inspecting the file, rather than plainly looking at the extension like Finder does)
  - `open -a Preview <>` to open a png file that has no extension
  - `ls -l <path to symlink` to check the path a symlink points to

- Pending questions
  - What is the best approach re this "data duplication"

>   - Where is the content of the files stored? Is it in the GIN server?

Reviewer: The answer is yes (as of your experiments today), right?
And then this PR could "close" #79.
